packages/headless/src/features/folding/folding-slice.ts

```diff
@@ -37,70 +42,113 @@ function areDefinedAndEqual<T>(
   return (value1 || value2) !== undefined && value1 === value2;
 }
 
-function flattenResultTree(result: ResultWithFolding): ResultWithFolding[] {
-  const results = [result];
-  if (result.parentResult) {
-    results.unshift(result.parentResult);
-  }
-  if (result.childResults?.length) {
-    results.push(
-      ...result.childResults.flatMap((childResult) =>
-        flattenResultTree(childResult)
-      )
-    );
+function getFoldedResults(result: ResultWithFolding): ResultWithFolding[] {
+  if (!result.childResults) {
+    return [];
   }
-  return results;
+  return result.childResults.flatMap((childResult) => [
+    childResult,
+    ...getFoldedResults(childResult),
+  ]);
 }
 
-function foldResult(
-  source: ResultWithFolding,
+function addChildren(
+  parent: ResultWithFolding,
   results: ResultWithFolding[],
   fields: FoldingFields
 ): FoldedResult {
-  const sourceChildValue = getChildField(source, fields);
+  const sourceChildValue = getChildField(parent, fields);
   return {
-    ...source,
+    ...parent,
     children: sourceChildValue
       ? results
           .filter((result) => {
             const isSameResultAsSource =
-              getChildField(result, fields) === getChildField(source, fields);
+              getChildField(result, fields) === getChildField(parent, fields);
             const isChildOfSource =
               getParentField(result, fields) === sourceChildValue;
             return isChildOfSource && !isSameResultAsSource;
           })
-          .map((result) => foldResult(result, results, fields))
+          .map((result) => addChildren(result, results, fields))
       : [],
   };
 }
```

Comment:
Like above, I'm going to update the other PR with this change. `addChildren` and `parent` felt much more relevant.